### PR TITLE
rustdoc: intra doc links link to re-exported primitives

### LINKED ISSUE #77267: Intra-doc links don't work for associated items on re-exported primitives
I tried this code:

```rust
/// [std::primitive::i32::MAX]
pub fn f() {}
```

I expected to see this happen: Rustdoc links to https://doc.rust-lang.org/stable/std/primitive.i32.html#associatedconstant.MAX

Instead, this happened: Rustdoc gives an error:

```
warning: unresolved link to `std::primitive::i32::MAX`
 --> prim-assoc.rs:1:6
  |
1 | /// [std::primitive::i32::MAX]
  |      ^^^^^^^^^^^^^^^^^^^^^^^^ the builtin type `i32` has no associated item named `MAX`
  |
  = note: `#[warn(broken_intra_doc_links)]` on by default
```

The issue is https://github.com/rust-lang/rust/blob/71bdb8481760f7c6b8540d3dd176ed54a92c0f31/src/librustdoc/passes/collect_intra_doc_links.rs#L315 which does string comparison instead of going through resolve.

### Meta

`rustdoc --version`: `rustdoc 1.48.0-nightly (f68e08933 2020-09-19)`

### PR DESCRIPTION
**Pull request by Stupremee**
Resolves #77267 

r? @jyn514 @GuillaumeGomez

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Nice! Does it work for methods on `slice` too?

**Comment by Stupremee** [comment]
No, it doesn't, but I can test if I can make it work.

**Comment by GuillaumeGomez** [comment]
Oh... That was the sad hope haha.

**Comment by jyn514** [comment]
@Stupremee https://github.com/rust-lang/rust/issues/63351 is a much harder issue, please don't mix it into any other changes.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -1892,14 +1890,34 @@ const PRIMITIVES: &[(&str, Res)] = &[
...
+                .ok()?
+                .1;
+            PRIMITIVES
+                .iter()
+                .find(|(_, prim_res)| match (res, prim_res) {
+                    (Res::PrimTy(ty1), Res::PrimTy(ty2)) if ty1 == *ty2 => true,
+                    (_, _) => false,
+                })
+                .copied()
+        })
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1920]: This does not look right. Instead of changing `is_primitive` to go through rustc_resolve, can you change some of the callsites not to use `is_primitive`? The only place that should be doing string comparisons is the ambiguity check for modules vs. primitives.

See also https://github.com/rust-lang/rust/pull/77743#discussion_r502614669.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -909,63 +909,61 @@ impl LinkCollector<'_, '_> {
...
-            } else {
-                disambiguator = None;
-                &link
-            }
-            .trim();
 
-            if path_str.contains(|ch: char| !(ch.is_alphanumeric() || ch == ':' || ch == '_')) {
-                return None;
-            }
+        let mut path_str = if let Ok((d, path)) = Disambiguator::from_str(&link) {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:914]: Can you split this refactor into a separate commit so the logic changes are easier to see?

**Comment by jyn514** [comment]
I'm going to close this in favor of https://github.com/rust-lang/rust/pull/77743, I like the approach there better. Sorry your code didn't make it in - if you're interested in working on other intra-doc issues let me know and I'm happy to mentor!
